src/main/java/org/candlepin/model/AbstractHibernateCurator.java

```diff
@@ -255,17 +258,47 @@ public E create(E entity, boolean flush) {
     /**
      * @return all entities for a particular type.
      */
-    public CandlepinQuery<E> listAll() {
-        DetachedCriteria criteria = this.createSecureDetachedCriteria();
+    public List<E> listAll() {
+        EntityManager em = this.getEntityManager();
+        CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();
+        CriteriaQuery<E> criteriaQuery = criteriaBuilder.createQuery(this.entityType);
+        Root<E> root = criteriaQuery.from(this.entityType);
+        criteriaQuery.select(root);
+
+        Predicate securityPredicate = this.getSecurityPredicate(this.entityType, criteriaBuilder, root);
+        if (securityPredicate != null) {
+            criteriaQuery.where(securityPredicate);
+        }
 
-        return this.cpQueryFactory.buildQuery(this.currentSession(), criteria);
+        return em.createQuery(criteriaQuery).getResultList();
     }
 
-    public CandlepinQuery<E> listAllByIds(Collection<? extends Serializable> ids) {
-        DetachedCriteria criteria = this.createSecureDetachedCriteria()
-            .add(CPRestrictions.in("id", ids));
+    public List<E> listAllByIds(Collection<? extends Serializable> ids) {
+        EntityManager em = this.getEntityManager();
+        CriteriaBuilder criteriaBuilder = em.getCriteriaBuilder();
+        CriteriaQuery<E> criteriaQuery = criteriaBuilder.createQuery(this.entityType);
+        Root<E> root = criteriaQuery.from(this.entityType);
+        criteriaQuery.select(root);
+
+        ParameterExpression<Collection> idList = criteriaBuilder.parameter(Collection.class);
+        Predicate predicate = root.get("id").in(idList);
+
+        Predicate securityPredicate = this.getSecurityPredicate(this.entityType, criteriaBuilder, root);
+        if (securityPredicate != null) {
+            predicate = criteriaBuilder.and(predicate, securityPredicate);
+        }
+        criteriaQuery.where(predicate);
+        TypedQuery<E> query = em.createQuery(criteriaQuery);
+
+        // to ensure uniqueness of the members of the input collection
+        Set<Serializable> dedupe = new HashSet<>();
```

Comment:
This addresses the immediate concern with output duplication, but now we're doing this on every invocation, which isn't necessary. We could also reuse the `ids` variable here since we're not interested in any methods on the set class itself, nor do we need to worry about keeping the original in scope:

```
if (!(ids instanceof Set)) {
  ids = new HashSet<>(ids);
}
```

This ensures we only do the reallocation and deduplication in instances where we don't already have a set, and discards the now-useless original input set from the scope.